### PR TITLE
verdant/typedef-optional-references

### PR DESCRIPTION
**Pull request by kartikarcot**
# Overview 

This PR is part of a push to upgrade GTSAM to C++17, and remove boost dependencies from GTSAM. In this PR we want to update our factor interfaces and the factor implementations that use arguments of the type boost::optional<A&> to raw pointers of the type A*. We wish to do that because the semantics of raw pointers are the same as that of an optional to a reference. This is in accordance to google style guide’s recommendation for function input parameters, [link here](https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs).

# Details

## Retaining `evaluateError` interface to accept both pointers and matrices
This was achieved by some templating and conditional compilation in `NoiseModelFactorN` class. A side-effect of this is that when a user inherits NoiseModelFactorN and implements it, they have to do something like this-
`using Base::evaluateError;` for all of the evaluateError overloads to be visible for the end user.

## Replacing std::binds with lambdas
std::bind had a problem inferring which evaluateError overload to use through template deduction. I replaced it with lambdas instead.

##  `OptionalJacobian`
I have replaced `boost::optional<Matrix&>` in functions whose, input argument sizes are known, with OptionalJacobians. This enables us to use that function with both pointers and l-value references of matrices.

# TODO
- [x] Need to still replace `boost::optional<Matrix&>` in the following lines of code after discussing with Frank:
```
 gtsam/geometry/CameraSet.h
112:      boost::optional<Matrix&> E = boost::none) const {
141:      boost::optional<Matrix&> E = boost::none) const {

gtsam/base/tests/testOptionalJacobian.cpp
35:  boost::optional<Matrix&> optional(dynamic);

gtsam/slam/SmartFactorBase.h
207:      boost::optional<Matrix&> E = boost::none) const {
244:      boost::optional<Matrix&> E = boost::none) const {}

gtsam_unstable/slam/SmartStereoProjectionFactor.h
457:      boost::optional<Matrix&> E = boost::none) const override {
```
- [ ] Need a review on `Pose3Upright.cpp` 
- [ ]  Need a review on `timePose2.cpp` (Deleted a test which looked like a copy paste to me)

### CODE REVIEW

**Comment by dellaert** [comment]
Seems like python tests are failing?

**Comment by jlblancoc** [comment]
I guess you might have already discussed this in person and discarded it but just in case:

    using optional_ref_t=std::optional<std::reference_wrapper<T>>;

Would be pure C++, avoid null pointers, have a clear semantics, and probably be equally efficient (not tested yet in compiler explorer, though...). 
I think Google's style is fine, but it was defined before C++17

File: CMakeLists.txt
Comment (dellaert) [inline CMakeLists.txt:9]: @jlblancoc I think this the right way and place to do so, right?
@kartikarcot I will probably do this in a PR before this one, so this diff will disappear from this PR.

**Comment by kartikarcot** [comment]
@jlblancoc we did consider it but decided to defer to google style guidelines about optional input output parameters
> Generally, use std::optional to represent optional by-value inputs, and use a const pointer when the non-optional form would have used a reference. Use non-const pointers to represent optional outputs and optional input/output parameters.

[Google Style Guide](https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs)

File: CMakeLists.txt
Comment (kartikarcot) [reply CMakeLists.txt:9]: Ok will do that

Comment (dellaert) [reply CMakeLists.txt:9]: No, you don’t have to remove it. It will be in both PRs

Comment (dellaert) [reply CMakeLists.txt:9]: Will just disappear from diff the moment i merge the other one

File: python/gtsam/tests/test_DiscreteBayesNet.py
Comment (kartikarcot) [reply python/gtsam/tests/test_DiscreteBayesNet.py:133]: Pasting the stacktrace here for visibility
```
======================================================================
ERROR: test_fragment (test_DiscreteBayesNet.TestDiscreteBayesNet)
Test evaluate/sampling/optimizing for Asia fragment.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/gtsam/gtsam/python/gtsam/tests/test_DiscreteBayesNet.py", line 133, in test_fragment
    self.assertEqual(len(values), 5)
TypeError: __len__(): incompatible function arguments. The following argument types are supported:
    1. (self: Dict[int, int]) -> int

Invoked with: DiscreteValues{0: 0, 3: 0, 4: 0, 5: 0, 6: 0}

----------------------------------------------------------------------
```

File: CMakeLists.txt
Comment (ProfFan) [reply CMakeLists.txt:9]: `target_compile_features(${TARGET_NAME} PRIVATE cxx_std_17)` is the right way to do it I think.

Comment (kartikarcot) [reply CMakeLists.txt:9]: Got it

File: gtsam/nonlinear/NonlinearFactor.h
Comment (ProfFan) [inline gtsam/nonlinear/NonlinearFactor.h:53]: Why raw pointers? Isn't `std::optional<std::reference_wrapper<T>>` working?

@jlblancoc

**Review by ProfFan — CHANGES_REQUESTED** [review]
I don't think it's wise to replace these with raw pointers, as raw pointers do not encode lifetime and ownership information.

**Review by ProfFan — APPROVED** [review]
Change to approval in case you need to unblock fast.

File: CMakeLists.txt
Comment (dellaert) [reply CMakeLists.txt:9]: Not sure - if there is a special CMAKE flag, will it not do the above automatically (and tailored to different compilers, if needed) ?

File: python/gtsam/tests/test_DiscreteBayesNet.py
Comment (dellaert) [reply python/gtsam/tests/test_DiscreteBayesNet.py:133]: @ProfFan any ideas?

File: CMakeLists.txt
Comment (jlblancoc) [reply CMakeLists.txt:9]: > @jlblancoc I think this the right way and place to do so, right?

Yes. But I agree with @ProfFan in that I would prefer to set it as a compile feature, but I think it must be public, not private (otherwise, users not using c++17 will for sure have build errors). 
So, it would be: 

    target_compile_features(${TARGET_NAME} PUBLIC cxx_std_17)

for the target `gtsam`, and perhaps `gtsam_unstable` although the latter would be redundant since it inherits the public flags from `gtsam`.

PS: From my own testing, I found [this additional command](https://github.com/MOLAorg/mola-common/blob/5b2001fb9b0503a3e38508cf0c5839c17047dfdb/mola_cmake_functions.cmake#L117-L123) was also needed to enable C++17 on MSVC, but give it a try first with `target_compile_features()` alone since it probably was a buggy cmake/msvc version by ~2020.

File: gtsam/nonlinear/NonlinearFactor.h
Comment (jlblancoc) [reply gtsam/nonlinear/NonlinearFactor.h:53]: It's up to you all, library API designers, both policies may make sense, despite my personal preference for not using pointers.

PS: Ownership and lifetime is neither ensured with the `std::optional<...>` type, btw... The only way to enforce and clarify both are (AFIK) smart pointers, so, dynamic memory, killing performance... I guess you must rely on well-designed code, validate with sanitizers, etc. Just as done so far :-)

File: CMakeLists.txt
Comment (dellaert) [reply CMakeLists.txt:9]: What is $TARGET_NAME supposed to be here? `gtsam` ?

Comment (ProfFan) [reply CMakeLists.txt:9]: I think so

File: doc/Code/LocalizationFactor.cpp
Comment (ProfFan) [inline doc/Code/LocalizationFactor.cpp:7]: Why is this necessary?

File: CMakeLists.txt
Comment (jlblancoc) [reply CMakeLists.txt:9]: Yes.

Comment (jlblancoc) [reply CMakeLists.txt:9]: Ah, of course then that line wouldn't be at the root cmake file anymore, but after defining the gtsam library target.

File: python/gtsam/tests/test_DiscreteBayesNet.py
Comment (ProfFan) [reply python/gtsam/tests/test_DiscreteBayesNet.py:133]: Probably an ODR violation. Need to check if there includes of `<pybind11/stl.h>` and whether `std::map<Key, size_t>` got accidentally wrapped automatically.

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [inline gtsam/nonlinear/NonlinearFactor.h:687]: May or may not be worth doing, but I think the pointer versions of these can be replaced with default args.  i.e.

```c++
template <typename... OptionalJacArgs, typename = IndexIsValid<sizeof...(OptionalJacArgs) + 1>>
virtual Vector evaluateError(const ValueTypes&... x, OptionalMatrixTypeT H1 = 0, OptionalJacArgs*... Hs) = 0;
```
or something along those lines (untested).  Or maybe I'm wrong, because otherwise I'm not sure why we did these weird overloading tricks previously.  Maybe I didn't realize that if you provide one default arg, the remaining template params automatically get default args.

File: gtsam/nonlinear/PriorFactor.h
Comment (gchenfc) [inline gtsam/nonlinear/PriorFactor.h:36]: Just a comment that I don't think this is so bad, because usually this won't be called directly anyway (but rather from a `NonlinearFactorGraph`) except in unit tests, and even then it's usually either (a) without the output args or (b) using factorTesting macros, so I don't think it will actually affect too much old code.

File: gtsam/nonlinear/PriorFactor.h
Comment (gchenfc) [inline gtsam/nonlinear/PriorFactor.h:98]: Just for reference, is it fair to say that this (and a lot of other instances) could be migrated by doing global find-replace: `boost::optional<Matrix&>` -> `OptionalMatrixType` ?

File: gtsam/slam/EssentialMatrixConstraint.cpp
Comment (gchenfc) [inline gtsam/slam/EssentialMatrixConstraint.cpp:59]: Why did this have to change?  Could this stay as
```c++
EssentialMatrix hx = EssentialMatrix::FromPose3(_1P2_, (Hp1 || Hp2) ? &D_hx_1P2 : 0);
```
?

File: gtsam/slam/EssentialMatrixFactor.h
Comment (gchenfc) [inline gtsam/slam/EssentialMatrixFactor.h:311]: I'm not sure I understand this comment - I think there might be a typo ("since the only overloaded function ... uses the type") and some missing comma's, and I'm not clear what "this" and "that" is referring to.  Sorry, not trying to be aggressive, just trying to give constructive feedback on how I interpreted when reading. :)

I am indeed curious because I would have expected the entire if-else block to be replaced by something like

```c++
Vector evaluateError( /* args */) {
  Matrix D_e_cameraE, D_cameraE_E; // 2*5, 5*5

  EssentialMatrix cameraE = E.rotate(cRb_, DE ? &D_cameraE_E : 0);
  Vector e = Base::evaluateError(cameraE, d, DE ? &D_e_cameraE : 0, Dd);
  if (DE) *DE = D_e_cameraE * D_cameraE_E;
  return e;
}
```

File: gtsam_unstable/slam/tests/testTSAMFactors.cpp
Comment (gchenfc) [inline gtsam_unstable/slam/tests/testTSAMFactors.cpp:137]: FYI for future reference I think it works if we do `boost::none` -> `(OptionalMatrixType*) 0` (or static_cast), but lambda looks better anyway :)

**Review by gchenfc — COMMENTED** [review]
Didn't read through entire thing in detail yet, but aside from the mentioned backward compatibility stuff, lgtm :)

**Comment by gchenfc** [comment]
> > using optional_ref_t=std::optional<std::reference_wrapper>;
> 
> That's interesting, I forgot about this possibility, but I actually like/prefer the `&` at the call site, to make it clear it's an output argument. We have a plan to switch to OptionalJacobian, to support fixed-size Jacobians, and will see whether we can use it there.

Is there a reason this PR can't already start using `OptionalJacobian` instead of `Matrix*` ?  Mainly for google style guide to use pointers for output arguments?

File: doc/Code/LocalizationFactor.cpp
Comment (kartikarcot) [reply doc/Code/LocalizationFactor.cpp:7]: This is to help compiler find the right overload of `evaluateError`. Since a function with the same name is overridden in the derived class (the pointer version), the compiler function name lookup logic just stops in the derived class and complains that it could not find the right overload. By declaring that we use other signatures from the parent class, we are able to access the matrix reference style of `evaluateError`, defined in the parent class, as well.

File: gtsam/nonlinear/NonlinearFactor.h
Comment (kartikarcot) [reply gtsam/nonlinear/NonlinearFactor.h:687]: Any remaining optional arguments are added by the line 686, recursively, right?
```
return evaluateError(x..., 
        std::forward<OptionalJacArgs>(H)..., static_cast<OptionalMatrixType>(OptionalNone));
```

File: gtsam/nonlinear/PriorFactor.h
Comment (kartikarcot) [reply gtsam/nonlinear/PriorFactor.h:98]: Yea that's right as long as the the operator overloads for `*` and `->` have the same semantics, if I got that right

File: gtsam/slam/EssentialMatrixConstraint.cpp
Comment (kartikarcot) [reply gtsam/slam/EssentialMatrixConstraint.cpp:59]: Yes it can, I think I changed it because I found it confusing. I will revert it.

File: gtsam/slam/EssentialMatrixFactor.h
Comment (kartikarcot) [reply gtsam/slam/EssentialMatrixFactor.h:311]: I have made the comment more clearer. Also by rephrasing it like this would mean cameraE will be computed even if DE is `nullptr`

File: gtsam_unstable/slam/tests/testTSAMFactors.cpp
Comment (kartikarcot) [reply gtsam_unstable/slam/tests/testTSAMFactors.cpp:137]: Ah I see

File: gtsam/slam/EssentialMatrixConstraint.cpp
Comment (kartikarcot) [reply gtsam/slam/EssentialMatrixConstraint.cpp:59]: Ah I changed it because `OptionalNone` and `D_hx_1P2` are of different types.

Comment (dellaert) [reply gtsam/slam/EssentialMatrixConstraint.cpp:59]: Well, could you instead use nullptr?

**Comment by dellaert** [comment]
> > > using optional_ref_t=std::optionalstd::reference_wrapper;
> > 
> > 
> > That's interesting, I forgot about this possibility, but I actually like/prefer the `&` at the call site, to make it clear it's an output argument. We have a plan to switch to OptionalJacobian, to support fixed-size Jacobians, and will see whether we can use it there.
> 
> Is there a reason this PR can't already start using `OptionalJacobian` instead of `Matrix*` ? Mainly for google style guide to use pointers for output arguments?

I might migrate to OptionalJacobian in a future PR. This one is the first cut at getting rid of boost::optional.

File: gtsam/slam/EssentialMatrixConstraint.cpp
Comment (kartikarcot) [reply gtsam/slam/EssentialMatrixConstraint.cpp:59]: No nullptr and D_hx_1P2 (matrix) are different types too. They can’t be used with a ternary operator. OptionalNone typecasts to nullptr

Comment (dellaert) [reply gtsam/slam/EssentialMatrixConstraint.cpp:59]: Check the original code, it’s a pointer. This idiom is used all over GTSAM so I’m a bit worried it’s broken.

Comment (dellaert) [reply gtsam/slam/EssentialMatrixConstraint.cpp:59]: I'll resolve - might fix in follow up PR

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h:687]: Ya sorry I wasn't very clear, I meant to say that I think it might be more readable with the default arguments rather than the lots-of-templates version.  But nbd :)

File: gtsam/navigation/tests/testAttitudeFactor.cpp
Comment (varunagrawal) [inline gtsam/navigation/tests/testAttitudeFactor.cpp:53]: This seems messy just to be a bit more aligned to the Google style guide...

Comment (gchenfc) [reply gtsam/navigation/tests/testAttitudeFactor.cpp:53]: I suspect this was just an oversight and that the old syntax could be used (and not a change driven by google style).  Although, personally, I actually find the lambda syntax easier to read than std::bind, but that's personal preference.

https://github.com/borglab/gtsam/pull/1392#discussion_r1083141328

**Pull request merged**
